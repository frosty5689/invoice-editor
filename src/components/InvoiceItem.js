import React from 'react';
import { connect } from 'react-redux';
import { deleteLineItem, updateLineItem } from '../redux/actions';

class InvoiceItem extends React.Component {
  render() {
    const item = this.props.item;
    const content = item.content;

    return (
      <tr className="Invoice-item">
        <td><input name="name" type="text" value={content.name} onChange={(e) => this.handleItemChange(e, item)}></input></td>
        <td><input name="quantity" type="number" min="1" value={content.quantity} onChange={(e) => this.handleItemChange(e, item)}></input></td>
        <td>$<input name="price" type="number" min="0.01" step="0.01" value={toCurrency(content.price)} onChange={(e) => this.handleItemChange(e, item)}></input></td>
        <td>${toCurrency(content.price * content.quantity)}</td>
        <td><button onClick={() => this.props.deleteLineItem(item.id)}>x</button></td>
      </tr>
    );
  }

  handleItemChange(event, item) {
    const target = event.target;
    const name = target.name;
    const newValue = target.type === "number" ? Number(target.value) : target.value;
    const id = item.id;
    const content = Object.assign({}, item.content);
    content[name] = newValue;
    this.props.updateLineItem(id, content);
  }
}



function toCurrency(number) {
  return number.toFixed(2);
}

export default connect(
  null,
  { deleteLineItem, updateLineItem }
)(InvoiceItem);