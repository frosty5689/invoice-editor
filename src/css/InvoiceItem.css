.Invoice-delete-item {
  border: 0;
  padding: 0;
  background-color: transparent;
  color: gray;
  cursor: pointer;
}

.Invoice-delete-item:hover {
  color: red;
}

.Invoice-delete-item:focus {
  outline: 0;
}