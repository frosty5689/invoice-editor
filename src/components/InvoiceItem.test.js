import React from 'react';
import { InvoiceItem } from './InvoiceItem';
import renderer from 'react-test-renderer';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-15';

Enzyme.configure({ adapter: new Adapter() });

describe('InvoiceItem component', () => {

  it('InvoiceItem should render correctly', () => {
    const item = {
      id: 1,
      content: {
        name: 'Widget',
        quantity: 2,
        price: 10.22
      }
    }
    const component = renderer.create(
      <InvoiceItem key={item.id} item={item} />,
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('InvoiceItem delete calls deleteLineItemFn', () => {
    const item = {
      id: 1,
      content: {
        name: 'Widget',
        quantity: 2,
        price: 10.22
      }
    }

    const mockDeleteLineItemFn = jest.fn();
    const component = shallow(
      <InvoiceItem key={item.id} item={item} deleteLineItem={mockDeleteLineItemFn} />,
    );

    component.find('button').simulate('click');
    expect(mockDeleteLineItemFn.mock.calls.length).toBe(1);
    expect(mockDeleteLineItemFn.mock.calls[0][0]).toBe(item.id);
  });

  it('InvoiceItem changing name calls updateLineItemFn', () => {
    const item = {
      id: 1,
      content: {
        name: 'Widget',
        quantity: 2,
        price: 10.22
      }
    }

    const event = { target: { name: 'name', value: 'Cogs' } };

    const mockUpdateLineItemFn = jest.fn();
    const component = shallow(
      <InvoiceItem key={item.id} item={item} updateLineItem={mockUpdateLineItemFn} />,
    );

    component.find('[name="name"]').simulate('change', event);

    expect(mockUpdateLineItemFn.mock.calls.length).toBe(1);
    expect(mockUpdateLineItemFn.mock.calls[0][0]).toBe(item.id);
    expect(mockUpdateLineItemFn.mock.calls[0][1].name).toBe(event.target.value);
  });

  it('InvoiceItem changing quantity calls updateLineItemFn', () => {
    const item = {
      id: 1,
      content: {
        name: 'Widget',
        quantity: 2,
        price: 10.22
      }
    }

    const event = { target: { name: 'quantity', value: 4 } };

    const mockUpdateLineItemFn = jest.fn();
    const component = shallow(
      <InvoiceItem key={item.id} item={item} updateLineItem={mockUpdateLineItemFn} />,
    );

    component.find('[name="quantity"]').simulate('change', event);

    expect(mockUpdateLineItemFn.mock.calls.length).toBe(1);
    expect(mockUpdateLineItemFn.mock.calls[0][0]).toBe(item.id);
    expect(mockUpdateLineItemFn.mock.calls[0][1].quantity).toBe(event.target.value);
  });

  it('InvoiceItem changing price calls updateLineItemFn', () => {
    const item = {
      id: 1,
      content: {
        name: 'Widget',
        quantity: 2,
        price: 10.22
      }
    }

    const event = { target: { name: 'price', value: 6.22 } };

    const mockUpdateLineItemFn = jest.fn();
    const component = shallow(
      <InvoiceItem key={item.id} item={item} updateLineItem={mockUpdateLineItemFn} />,
    );

    component.find('[name="price"]').simulate('change', event);

    expect(mockUpdateLineItemFn.mock.calls.length).toBe(1);
    expect(mockUpdateLineItemFn.mock.calls[0][0]).toBe(item.id);
    expect(mockUpdateLineItemFn.mock.calls[0][1].price).toBe(event.target.value);
  });

  it('InvoiceItem toCurrency should return two decimal places', () => {
    const currency = InvoiceItem.prototype.toCurrency(100);
    expect(currency).toBe('100.00');
  });
});