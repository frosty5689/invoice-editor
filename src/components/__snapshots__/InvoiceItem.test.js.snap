// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InvoiceItem component InvoiceItem should render correctly 1`] = `
<tr
  className="Invoice-item"
>
  <td>
    <input
      name="name"
      onChange={[Function]}
      type="text"
      value="Widget"
    />
  </td>
  <td>
    <input
      min="1"
      name="quantity"
      onChange={[Function]}
      type="number"
      value={2}
    />
  </td>
  <td>
    $
    <input
      min="0.01"
      name="price"
      onChange={[Function]}
      step="0.01"
      type="number"
      value="10.22"
    />
  </td>
  <td>
    $
    20.44
  </td>
  <td>
    <button
      onClick={[Function]}
    >
      x
    </button>
  </td>
</tr>
`;
